pkg/util/iptree/iptree.go

```diff
@@ -0,0 +1,602 @@
+		bitPosition = n.prefix.Bits()
+
+		// check we are in the right branch comparing the suffixes
+		if !n.prefix.Contains(address) {
+			break
+		}
+
+	}
+	var zeroT T
+	return zeroT, false
+}
+
+// LongestPrefixMatch returns the longest prefix match, the stored value and true if exist.
+func (t *Tree[T]) LongestPrefixMatch(subnet string) (string, T, bool) {
+	prefix, err := netip.ParsePrefix(subnet)
+	if err != nil {
+		var zeroT T
+		return "", zeroT, false
+	}
+	p, v, ok := t.longestPrefix(prefix)
+	if !ok {
+		var zeroT T
+		return "", zeroT, false
+	}
+	return p.String(), v, ok
+}
+
+func (t *Tree[T]) longestPrefix(prefix netip.Prefix) (netip.Prefix, T, bool) {
+	if prefix.Addr().Is6() != t.is6 {
+		var zeroT T
+		return netip.Prefix{}, zeroT, false
+	}
+
+	var last *node[T]
+	n := t.root
+	// bit position is given by the mask bits
+	bitPosition := 0
+	// mask the address
+	address := prefix.Masked().Addr()
+	mask := prefix.Bits()
+	for bitPosition <= mask {
+		if n.isLeaf() {
+			last = n
+		}
+		// we branch at the mask bit
+		if bitPosition == mask {
+			break
+		}
+
+		// Look for a child checking the bit position after the mask
+		n = n.child[getBitFromAddr(address, bitPosition+1)]
+		if n == nil {
+			break
+		}
+
+		// update the new bit position with the new node mask
+		bitPosition = n.prefix.Bits()
+
+		// check we are in the right branch
+		if !n.prefix.Contains(address) {
+			break
+		}
+
+	}
+	if last != nil {
+		return last.prefix, last.val, true
+	}
+	var zeroT T
+	return netip.Prefix{}, zeroT, false
+
+}
+
+// ShortestPrefixMatch returns the shortest prefix match, the stored value and true if exist.
+func (t *Tree[T]) ShortestPrefixMatch(subnet string) (string, T, bool) {
+	prefix, err := netip.ParsePrefix(subnet)
+	if err != nil {
+		var zeroT T
+		return "", zeroT, false
+	}
+	p, v, ok := t.shortestPrefix(prefix)
+	if !ok {
+		var zeroT T
+		return "", zeroT, false
+	}
+	return p.String(), v, ok
+}
+
+func (t *Tree[T]) shortestPrefix(prefix netip.Prefix) (netip.Prefix, T, bool) {
+	if prefix.Addr().Is6() != t.is6 {
+		var zeroT T
+		return netip.Prefix{}, zeroT, false
+	}
+
+	n := t.root
+	// bit position is given by the mask bits
+	bitPosition := 0
+	// mask the address
+	address := prefix.Masked().Addr()
+	mask := prefix.Bits()
+	for bitPosition <= mask {
+		if n.isLeaf() {
+			return n.prefix, n.val, true
+		}
+		// we branch at the mask bit
+		if bitPosition == mask {
+			break
+		}
+
+		// Look for a child checking the bit position after the mask
+		n = n.child[getBitFromAddr(address, bitPosition+1)]
+		if n == nil {
+			break
+		}
+
+		// update the new bit position with the new node mask
+		bitPosition = n.prefix.Bits()
+
+		// check we are in the right branch
+		if !n.prefix.Contains(address) {
+			break
+		}
+
+	}
+	var zeroT T
+	return netip.Prefix{}, zeroT, false
+
+}
+
+// Insert the specified subnet and the value associated.
+// Retyrns true if updated.
+func (t *Tree[T]) Insert(subnet string, v T) (T, bool) {
```

Comment:
it borrows the semantics from https://github.com/armon/go-radix, the tree implementation is based on that one

Reply:
That implementation always returns either `old, true` or `nil, false`, which makes more sense. It doesn't have the hard-to-explain `v, false` case.

Reply:
I can make the return value a pointer *T

Reply:
No, it's not about that. I mean that line 301 should be `return zeroT, false` rather than `return v, false`. Or if not, then you _really_ need to update the documentation to explain why it makes sense to return `v` rather than `zeroT` only in that one case.

Reply:
I think that returning the value does not make much sense and is hard to think about in this case, will change the API to just return true if update